oda_api/token.py

```diff
@@ -62,57 +69,125 @@ def decode_oda_token(token: str, secret_key=None, allow_invalid=False) -> dict:
 
     raise RuntimeError()
 
+
 def decode_oauth2_token(token: str):
     # usually comes in cookies['_oauth2_proxy']
     return json.loads(base64.b64decode(token.split(".")[0]+"=").decode())
 
-#TODO: move to dynaconf
-def discover_token(
+
+def get_token_roles(decoded_token):
+    # extract role(s)
+    roles = None
+    if isinstance(decoded_token['roles'], str):
+        roles = decoded_token['roles'].split(',') if 'roles' in decoded_token else []
+        roles[:] = [r.strip() for r in roles]
+    elif isinstance(decoded_token['roles'], list):
+        roles = decoded_token['roles'] if 'roles' in decoded_token else []
+        roles[:] = [r.strip() for r in roles]
+    return roles
+
+
+def rewrite_token(new_token,
+                  token_write_method: TokenLocation = None
+                  ):
+    current_token, discover_method = discover_token_and_method(allow_invalid=True)
+    current_decoded_token = decode_oda_token(current_token, allow_invalid=True)
+    new_decoded_token = decode_oda_token(new_token, allow_invalid=True)
+
+    current_decoded_token_expires_in_s = current_decoded_token['exp'] - time.time()
+    new_decoded_token_expires_in_s = new_decoded_token['exp'] - time.time()
+
+    if new_decoded_token_expires_in_s < current_decoded_token_expires_in_s:
+        logger.info("the new token has will expire before the current one")
+
+    current_decoded_token_roles = get_token_roles(current_decoded_token)
+    new_decoded_token_roles = get_token_roles(new_decoded_token)
+    if set(current_decoded_token_roles) - set(new_decoded_token_roles) != set():
+        logger.info("The new token has a different set of roles than the current one:\n"
+                    f"roles current token: {current_decoded_token_roles}\n"
+                    f"roles new token: {new_decoded_token_roles}")
+
+    if token_write_method is not None:
+        # TODO perhaps use a more appropriate location
+        with open("old-oda-token_" + str(time.time()), 'w') as ft:
+            ft.write(current_token)
+        if discover_method is not None:
+            if discover_method == TokenLocation.ODA_ENV_VAR:
```

Comment:
this exterminates all tokens and can be quite dangerous. What if some of them were better. I have it happen sometimes: I may be using underprivileged user, but I have high-priveged token stored.

I suggest to not silently overwrite. See next comment.

Reply:
> this exterminates all tokens and can be quite dangerous. What if some of them were better. I have it happen sometimes: I may be using underprivileged user, but I have high-priveged token stored.
> 

In such a situation (old token better than the new one) there is the check on the roles, and that partally verifies exactly this

Reply:
Right, good point, we need to check that all other token settings are the same. Refresh should only change the duration - that's its meaning. Maybe roles in a rare case.

Reply:
Anyway, I still think this is way too dangerous and in case of any bug causes. Maybe this can be enabled with an extra option `discard_all_existing_tokens` which should be off by default.

Reply:
`discard_all_existing_tokens` is now an argument of the function, False by default.